src/Disks/ObjectStorages/MetadataStorageFromDiskTransactionOperations.cpp

```diff
@@ -146,20 +147,31 @@ void RemoveRecursiveOperation::finalize()
         disk.removeRecursive(path);
 }
 
-CreateHardlinkOperation::CreateHardlinkOperation(const std::string & path_from_, const std::string & path_to_, IDisk & disk_)
+CreateHardlinkOperation::CreateHardlinkOperation(const std::string & path_from_, const std::string & path_to_, IDisk & disk_, const MetadataStorageFromDisk & metadata_storage_)
     : path_from(path_from_)
     , path_to(path_to_)
     , disk(disk_)
+    , metadata_storage(metadata_storage_)
 {
 }
 
-void CreateHardlinkOperation::execute()
+void CreateHardlinkOperation::execute(std::unique_lock<std::shared_mutex> & lock)
 {
+    auto metadata = metadata_storage.readMetadataUnlocked(path_from, lock);
+
+    metadata->incrementRefCount();
```

Comment:
Was the issue because of the previous version didn't increment the ref count?

Reply:
No previous version did it without lock.

Reply:
No, it call increment before, but without lock (out of execute scope)